Data/Row/Variants.hs

```diff
@@ -408,6 +411,112 @@ fromLabels mk = getCompose $ metamorph @_ @ρ @c @FlipConst @(Const ()) @(Compos
         doCons l (FlipConst (Compose v)) = Compose $ IsJust l <$> mk l <|> extend @τ l <$> v
           \\ extendHas @ρ @ℓ @τ
 
+{--------------------------------------------------------------------
+  Functions for variants of ApSingle
+--------------------------------------------------------------------}
+
+newtype VApS x (fs :: Row (* -> *)) = VApS { unVApS :: Var (ApSingle fs x) }
+newtype FlipApp (x :: *) (f :: * -> *) = FlipApp (f x)
+
+eraseSingle
+  :: forall (c :: (* -> *) -> Constraint) (fs :: Row (* -> *)) (x :: *) (y :: *)
+   . (Forall fs c)
+  => (forall f . (c f) => f x -> y)
+  -> Var (ApSingle fs x)
+  -> y
+eraseSingle f =
+  getConst
+    . metamorph @_ @fs @c @Either @(VApS x) @(Const y) @(FlipApp x)
+        Proxy
+        doNil
+        doUncons
+        doCons
+    . VApS
+ where
+  doNil = impossible . unsafeCoerce . unVApS
+
+  doUncons
+    :: forall l f fs
+      . ( c f
+        , fs .! l ≈ f
+        , KnownSymbol l
+        )
+    => Label l
+    -> VApS x fs
+    -> Either (FlipApp x f) (VApS x (fs .- l))
+  doUncons l = (FlipApp +++ VApS) . (flip trial l \\ apSingleHas @fs @l @f @x) . unVApS
+
+  doCons
+    :: forall l f fs
+     . (c f)
+    => Label l
+    -> Either (FlipApp x f) (Const y fs)
+    -> Const y (Extend l f fs)
+  doCons _ (Left (FlipApp v)) = Const (f v)
+  doCons _ (Right (Const  y)) = Const y
+
+mapSingle
+  :: forall (c :: (* -> *) -> Constraint) (fs :: Row (* -> *)) (x :: *) (y :: *)
+   . (Forall fs c)
+  => (forall f . (c f) => f x -> f y)
+  -> Var (ApSingle fs x)
+  -> Var (ApSingle fs y)
+mapSingle f = unVApS . metamorph @_ @fs @c @Either @(VApS x) @(VApS y) @(FlipApp x)
+             Proxy doNil doUncons doCons . VApS
+ where
+  doNil = impossible . unsafeCoerce . unVApS
+
+  doUncons :: forall l f fs
+           .  ( c f, fs .! l ≈ f, KnownSymbol l)
+           => Label l -> VApS x fs -> Either (FlipApp x f) (VApS x (fs .- l))
+  doUncons l = (FlipApp +++ VApS)
+             . flip (trial \\ apSingleHas @fs @l @f @x) l
+             . unVApS
+
+  doCons :: forall l f fs. (KnownSymbol l, c f)
+         => Label l
+         -> Either (FlipApp x f) (VApS y fs)
+         -> VApS y (Extend l f fs)
+  doCons (toKey -> l) (Left (FlipApp x)) = VApS . OneOf l . HideType $ f x
+  doCons l (Right (VApS v)) = VApS $
+    extend @(f y) l v
+      \\ apSingleExtendSwap @_ @l @y @fs @f
+
+eraseZipSingle :: forall c fs (x :: *) (y :: *) z
+                . (Forall fs c)
+               => (forall f. c f => Either (f x, f y) (Either (Text, f x) (Text, f y)) -> z)
+               -> Var (ApSingle fs x) -> Var (ApSingle fs y) -> z
```

Comment:
This type seems a little crazy.  In the case that the variant isn't at the same label, I understand the desire to know which one "comes first" — I do the same thing in the `Ord` instance for `Var` at the top of this module — but I'm not sure it's a good call for a public function like this.  I wonder if there's a more general `eraseZip` that I've been missing too, like where the given function is of type like:
```haskell
forall a b. c a => Either (Label, a, a) ((Label, a), (Label, b)) -> c
```
With this `generalEraseZip`, we could recover eraseZip pretty easily:
```haskell
eraseZip f = generalEraseZip $ \case
  Left (_, a, b) -> Just (f a b)
  Right _ -> Nothing
```
And then we could also use it in the `Ord` instance by doing:
```haskell
compare = generalEraseZip $ \case
  Left (_, a, b) -> compare a b
  Right ((la,_),(lb,_)) -> compare (show la, show lb)
```
One really nice thing about this is that it's safe to reordering, meaning that if we change row-types in the future so that rows aren't necessarily ordered alphabetically be label, the meaning of `compare` won't suddenly change.

Anyway, I think a similar thing can be done for `eraseZipSingle` (or possibly the `ApSingle` version could be written in terms of `generalEraseZip` via the `ActsOn` trick I talked about in a previous comment).